src/windows/network.rs

```diff
@@ -246,4 +287,137 @@ impl NetworkExt for NetworkData {
+                    self.unicast_address = (*address).Next;
+                    if self.unicast_address.is_null() {
+                        // if we have visited all unicast addresses, move to next adapter
+                        self.adapter = (*adapter).Next;
+                    }
+                    // FIXME: should we perform a null check on (*address).Address.lpSockaddr
+                    let sock_addr = (*address).Address.lpSockaddr;
+
+                    match (*sock_addr).sa_family as _ {
+                        AF_INET => {
+                            let sock_addr = sock_addr as *const SOCKADDR_IN;
+                            let sock_addr = (*sock_addr).sin_addr.S_un.S_addr();
+                            let mut subnet_mask = 0 as ULONG;
+                            // only avaialbe on vista and later
+                            // https://learn.microsoft.com/zh-cn/windows/win32/api/netioapi/nf-netioapi-convertlengthtoipv4mask
+                            ConvertLengthToIpv4Mask(
+                                (*address).OnLinkPrefixLength as _,
+                                &mut subnet_mask as _,
+                            );
+
+                            Some((
+                                interface_name,
+                                InterfaceAddress::IPv4(
+                                    Ipv4Addr::from(htonl(*sock_addr)),
+                                    Ipv4Addr::from(htonl(subnet_mask)),
+                                ),
+                            ))
+                        }
+                        _ => Some((interface_name, InterfaceAddress::NotImplemented)),
+                    }
+                }
+            } else {
+                // Not sure whether error can occur when parsing adapter name.
+                // If we met an error, move to the next adapter
+                self.adapter = (*adapter).Next;
+                self.next()
+            }
+        }
+    }
+}
+
+impl Drop for InterfaceAddressIterator {
+    fn drop(&mut self) {
+        unsafe {
+            libc::malloc(self.buf as _);
+        }
+    }
+}
+
+fn get_interface_address() -> Result<InterfaceAddressIterator, String> {
+    unsafe {
+        // https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getadaptersaddresses#remarks
+        // A 15k buffer is recommended
+        let mut size: u32 = 16 * 1024;
+        let buf = libc::malloc(size as usize) as PIP_ADAPTER_ADDRESSES;
+        if buf.is_null() {
+            // TODO: more details
+            return Err("malloc failed".to_string());
+        }
+
+        let ret = GetAdaptersAddresses(
+            AF_UNSPEC as u32,
+            GAA_FLAG_SKIP_MULTICAST | GAA_FLAG_SKIP_ANYCAST | GAA_FLAG_SKIP_DNS_SERVER,
+            NULL,
+            buf,
+            &mut size,
+        );
+
+        if ret != ERROR_SUCCESS {
+            return Err("GetAdaptersAddresses() failed".to_string());
```

Comment:
Please also add `GetLastError()` returned value. One important point though: in winapi [docs](https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getadaptersaddresses), they seem to retry to retrieve this information three times maximum. Not sure how relevant this is though...

Reply:
According to MSDN, GetAdaptersAddresses should be called twice. At the first time we obtain the actual size required and at the second time we obtain the adapter information.
It is placed in src/network_helper_win.rs now.